x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/repositories/GetCcrRestoreFileChunkAction.java

```diff
@@ -0,0 +1,116 @@
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.lease.Releasable;
+import org.elasticsearch.common.util.concurrent.AbstractRunnable;
+import org.elasticsearch.index.engine.Engine;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportActionProxy;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.ccr.repository.CcrRestoreSourceService;
+
+import java.io.IOException;
+
+public class GetCcrRestoreFileChunkAction extends Action<GetCcrRestoreFileChunkAction.GetCcrRestoreFileChunkResponse> {
+
+    public static final GetCcrRestoreFileChunkAction INSTANCE = new GetCcrRestoreFileChunkAction();
+    public static final String NAME = "internal:admin/ccr/restore/file_chunk/get";
+
+    private GetCcrRestoreFileChunkAction() {
+        super(NAME);
+    }
+
+    @Override
+    public GetCcrRestoreFileChunkResponse newResponse() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Writeable.Reader<GetCcrRestoreFileChunkResponse> getResponseReader() {
+        return GetCcrRestoreFileChunkResponse::new;
+    }
+
+
+    public static class TransportGetCcrRestoreFileChunkAction
+        extends HandledTransportAction<GetCcrRestoreFileChunkRequest, GetCcrRestoreFileChunkAction.GetCcrRestoreFileChunkResponse> {
+
+        private final CcrRestoreSourceService restoreSourceService;
+        private final ThreadPool threadPool;
+
+        @Inject
+        public TransportGetCcrRestoreFileChunkAction(ThreadPool threadPool, TransportService transportService, ActionFilters actionFilters,
+                                                     CcrRestoreSourceService restoreSourceService) {
+            super(NAME, transportService, actionFilters, GetCcrRestoreFileChunkRequest::new);
+            TransportActionProxy.registerProxyAction(transportService, NAME, GetCcrRestoreFileChunkResponse::new);
+            this.threadPool = threadPool;
+            this.restoreSourceService = restoreSourceService;
+        }
+
+        @Override
+        protected void doExecute(Task task, GetCcrRestoreFileChunkRequest request,
+                                 ActionListener<GetCcrRestoreFileChunkResponse> listener) {
+            threadPool.generic().execute(new AbstractRunnable() {
+                @Override
+                public void onFailure(Exception e) {
+                    listener.onFailure(e);
+                }
+
+                @Override
+                protected void doRun() throws Exception {
+                    CcrRestoreSourceService.Session session = restoreSourceService.getSession(request.getSessionUUID());
+                    Engine.IndexCommitRef refCommit = session.refCommit();
+                    try (IndexInput in = refCommit.getIndexCommit().getDirectory().openInput(request.getFileName(), IOContext.READONCE);
```

Comment:
We can always keep more state around in the session on the sender side (e.g. the open IndexInput) to make this more efficient. We can make that change in this PR, or as a follow-up (as I think it will have little impact on the other bits in this PR).

Reply:
if we keep such a state with a file handle we have to make sure we release it. It might have all kinds of implications on recovery, file removal, testing etc. I think it's an essential part of this chnage that must be sorted out ahead of time.

Reply:
I made a change to cache a file until a new file is requested.